File: .leetcode/1005.k-次取反后最大化的数组和.java
import java.util.PriorityQueue;

/*
 * @lc app=leetcode.cn id=1005 lang=java
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
class Solution {
    public int largestSumAfterKNegations(int[] nums, int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        int n = nums.length;
        int sum = 0;
        for(int i = 0;i < n;i++){
            sum += nums[i];
            heap.add(nums[i]);
        }
        while(k-- > 0){
            int temp = heap.poll();
            temp = -temp;
            heap.add(temp);
            sum += 2*temp;
        }
        return sum;
    }
}
// @lc code=end

File: .leetcode/1005.k-次取反后最大化的数组和.cpp
/*
 * @lc app=leetcode.cn id=1005 lang=cpp
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int largestSumAfterKNegations(vector<int>& A, int K) {
        sort(A.begin(),A.end());
        int i = 0;
        for(i;A[i]<0;i++);
        int f = i;
        if(A[0]>=0)f=-1;
        int sum = 0;
        for(i;i<A.size();i++)sum+=A[i];
        if(K<f){
            for(i=0;i<=f-K;i++){
                sum+=-A[i];
            }
            for(i;i<f;i++)sum+=A[i];
            return sum;
        }else{
            if(f==-1){
                sum-=(K)%2?2*A[0]:0;
            }else{
            for(i = 0;i<f;i++)sum+=-A[i];
            sum+=(K-f)%2?A[f-1]:0;
            return sum;
            }
        }
        return sum;
    }
};
// @lc code=end

